### PR TITLE
Enhance Base Codegen to do clean up when generation fails

### PR DESCRIPTION
**Pull request by karthijrk**
When generator failed to generate code, base codegen will handle the clean up. It is important to remove those broken functions else llvm exit when it compiles 

@foyzur @hardikar @armenatzoglou @gcaragea Please take a look when you get chance.

### CODE REVIEW

File: src/backend/codegen/include/codegen/base_codegen.h [hunk 1]
```diff
@@ -43,6 +45,16 @@ class BaseCodegen: public CodegenInterface {
...
     is_generated_ = GenerateCodeInternal(codegen_utils);
+    if (!is_generated_) {
+      // If failed to generate, make sure we do clean up
+      // by erasing all the llvm functions.
+      for (llvm::Function* function : uncompiled_generated_functions_) {
+        assert(nullptr != function);
+        function->eraseFromParent();
+      }
+    }
+    // We don't need to keep these pointer any more
```
Comment (armenatzoglou) [inline src/backend/codegen/include/codegen/base_codegen.h]: typo: pointer**s**

File: src/backend/codegen/include/codegen/base_codegen.h [hunk 2]
```diff
@@ -147,12 +159,37 @@ class BaseCodegen: public CodegenInterface {
    **/
   virtual bool GenerateCodeInternal(gpcodegen::CodegenUtils* codegen_utils) = 0;
 
+  /**
+   * @brief Create llvm Function for given type and store the function pointer
+   *        in vector
+   *
+   * @note  If generation fails, this class takes responsible to clean up all
```
Comment (armenatzoglou) [inline src/backend/codegen/include/codegen/base_codegen.h]: typo: takes **responsibility**

**Comment by armenatzoglou** [comment]
Nice work! +1

File: src/backend/codegen/include/codegen/base_codegen.h (hunk 2 above)
Comment (karthijrk) [inline src/backend/codegen/include/codegen/base_codegen.h]: oh!! good catch

File: src/backend/codegen/ExecVariableList_codegen.cc [hunk 3]
```diff
@@ -159,9 +159,8 @@ bool ExecVariableListCodegen::GenerateExecVariableList(
   }
 
   // So looks like we're going to generate code
-  llvm::Function* ExecVariableList_func = codegen_utils->
-      CreateFunction<ExecVariableListFn>(
-          GetUniqueFuncName());
+  llvm::Function* ExecVariableList_func = CreateFunction<ExecVariableListFn>(
```
Comment (hardikar) [inline src/backend/codegen/ExecVariableList_codegen.cc]: Can you split these hunks as a separate commit, when pushing.

File: src/backend/codegen/include/codegen/base_codegen.h [hunk 4]
```diff
@@ -147,12 +159,37 @@ class BaseCodegen: public CodegenInterface {
...
+    return function;
+  }
+
  private:
   std::string orig_func_name_;
   std::string unique_func_name_;
   FuncPtrType regular_func_ptr_;
   FuncPtrType* ptr_to_chosen_func_ptr_;
   bool is_generated_;
+  std::vector<llvm::Function*> uncompiled_generated_functions_;
```
Comment (hardikar) [inline src/backend/codegen/include/codegen/base_codegen.h]: Add a comment please.

**Comment by hardikar** [comment]
Rest LGTM. Thanks for doing this! Let's just chat once again in the morning about this design. I believe you're right this is better than to put it in CodegenUtils.

File: src/backend/codegen/ExecVariableList_codegen.cc (hunk 3 above)
Comment (karthijrk) [inline src/backend/codegen/ExecVariableList_codegen.cc]: :)

**Comment by karthijrk** [comment]
Great discussion @hardikar. Let bring those shared module use case in one of tech review meeting.

File: src/backend/codegen/include/codegen/base_codegen.h [hunk 5]
```diff
@@ -147,12 +159,37 @@ class BaseCodegen: public CodegenInterface {
...
+   *        the functions it created
+   *
+   * @tparam FunctionType Type of the function to create
+   *
+   * @param codegen_utils Utility to ease the code generation process.
+   * @param function_name Name of the function to create
+   * @return llvm::Function pointer
+   **/
+  template <typename FunctionType>
+  llvm::Function* CreateFunction(gpcodegen::CodegenUtils* codegen_utils,
```
Comment (foyzur) [inline src/backend/codegen/include/codegen/base_codegen.h]: Nice wrapper to catch the list of generated function. This is unfortunately more error prone as it depends on programmer's careful calling from sub-classes. Ideally I would prefer either:

1. Central calling from base class using more template methods and tracking the generated functions
2. A walk of newly generated functions from a central location like LLVM

Unfortunately, neither is feasible for this problem. Therefore, this LGTM as the best available alternative.

**Comment by foyzur** [comment]
LGTM

File: src/backend/codegen/include/codegen/base_codegen.h (hunk 5 above)
Comment (karthijrk) [inline src/backend/codegen/include/codegen/base_codegen.h]: Yeah you are right. Programmer has to call basecodegen's create function. I can't think better way other than this. If we come up with better way we will definitely revisit and update this.
